presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushdownSubscripts.java

```diff
@@ -0,0 +1,302 @@
+                    node.getId(),
+                    node.getTable(),
+                    node.getOutputSymbols(),
+                    annotateColumnsWithSubfields(node.getAssignments()),
+                    node.getLayout(),
+                    node.getCurrentConstraint(),
+                    node.getEnforcedConstraint());
+        }
+
+        @Override
+        public PlanNode visitFilter(FilterNode node, RewriteContext<Void> context)
+        {
+            collectSubfieldPaths(node.getPredicate());
+
+            PlanNode source = context.rewrite(node.getSource(), context.get());
+            return node.replaceChildren(ImmutableList.of(source));
+        }
+
+        @Override
+        public PlanNode visitUnnest(UnnestNode node, RewriteContext<Void> context)
+        {
+            processUnnestPaths(node.getUnnestSymbols());
+
+            PlanNode source = context.rewrite(node.getSource(), context.get());
+            return node.replaceChildren(ImmutableList.of(source));
+        }
+
+        @Override
+        public PlanNode visitProject(ProjectNode node, RewriteContext<Void> context)
+        {
+            processProjectionPaths(node.getAssignments());
+
+            PlanNode source = context.rewrite(node.getSource(), context.get());
+            return node.replaceChildren(ImmutableList.of(source));
+        }
+
+        @Override
+        public PlanNode visitOutput(OutputNode node, RewriteContext<Void> context)
+        {
+            fullColumnUses.addAll(node.getOutputSymbols());
+
+            PlanNode source = context.rewrite(node.getSource(), context.get());
+            return node.replaceChildren(ImmutableList.of(source));
+        }
+
+        private Map<Symbol, ColumnHandle> annotateColumnsWithSubfields(Map<Symbol, ColumnHandle> assignments)
+        {
+            if (subfieldPaths.isEmpty()) {
+                return assignments;
+            }
+
+            Map<Symbol, ColumnHandle> newAssignments = new HashMap<>();
+            for (Map.Entry<Symbol, ColumnHandle> entry : assignments.entrySet()) {
+                if (fullColumnUses.contains(entry.getKey())) {
+                    newAssignments.put(entry.getKey(), entry.getValue());
+                    continue;
+                }
+
+                List<SubfieldPath> subfields = new ArrayList();
+                for (SubfieldPath path : subfieldPaths) {
+                    if (path.getColumnName().equals(entry.getKey().getName())) {
+                        subfields.add(path);
+                    }
+                }
+                if (subfields.isEmpty()) {
+                    newAssignments.put(entry.getKey(), entry.getValue());
+                    continue;
+                }
+                // Compute the leaf subfields. If we have a.b.c and
+                // a.b then a.b is the result. If a path is a prefix
+                // of another path, then the longer is discarded.
+                List<SubfieldPath> leafPaths = new ArrayList();
```

Comment:
I've read it a couple of times now, and I believe this comment isn't accurate.  I think it actually returns `a.b.c`, not `a.b`.

Reply:
@tdcmeehan "a.b".isPrefix("a.b.c") returns true, hence, "a.b" will be added to leafPaths, but "a.b.c" won't.